### DISCUSSION TITLE
Edge classification in GATConv

### DISCUSSION THREAD

**Discussion by haoyuan-l** [pyg-team/pytorch_geometric#5655] [Q&A] [answered]
Hello. I would like to do node&edge classification by GATConv. Currently, I have implemented a model to do the node classification, but I'm not sure if GATConv can be used for edge classification(is_in_path/is_not_in_path). I saw a solution in another post, and it suggests taking a dot product between nodes connected by edges and passing it to the BCE. I deployed a similar structure, but the model accuracy is relatively low. I'm not proficient in Pytorch Geometric, so any suggestions will be appreciated!      
```
            # x is the node matrix

            src, dst = edge_index 

            # pass to several GAT layers
            x = GATConv(x=x, edge_index=edge_index, edge_attr=edge_attr)

            score = (x[src] * x[dst ] * edge_attr).sum(dim=-1)

            x = F.dropout(x, p=0.5, training=self.training)

            score = torch.sigmoid(score)
            
            return x, score
```

**Comment by wsad1** [comment, marked as answer]
Check out this example on link prediction https://github.com/pyg-team/pytorch_geometric/blob/master/examples/link_pred.py.  Some things you could try if you're not already doing so is
1. add negative sampling
2. use `BCEWithLogitsLoss` for stability.

**Reply by haoyuan-l** [reply]
> Check out this example on link prediction https://github.com/pyg-team/pytorch_geometric/blob/master/examples/link_pred.py. Some things you could try if you're not already doing so is
> 
> 1. add negative sampling
> 2. use `BCEWithLogitsLoss` for stability.


Hello, thank you for your kind response. I have tried the BCEWithLogitsLoss, but the result is still unchanged. I'm not sure if negative sampling is suitable for my case. I've been working on shortest path prediction(it contains node&edge prediction), which means all edges within the path must be connected.

The node prediction is about 70% accuracy, and the edge prediction is even worse, less than 50%. I think this may be because the GAT can't return updated edge features. And it can only return `alpha `(attention weights for each edge) and `x` (node features). That means I used the same edge_attr for several consecutive GAT layers instead of  new edge_attr.

**Reply by wsad1** [reply]
If you want the layers to "learn" the `edge_attr`, you could pass the `alpha` returned by `GATConv` to the next layer like
```
x, alpha = GATConv(x, edge_index, edge_attr) 
x, alpha = GATConv(x, edge_index, alpha)
```

**Reply by haoyuan-l** [reply]
![image](https://user-images.githubusercontent.com/91633635/195690332-abf4345b-8de0-40c6-a801-9e6a20900019.png)
Hi, wsad. To learn the `edge_attr`, we just simply replace `edge_attr` with `alpha` and pass it to the next graph attention layer. Isn't the `alpha` is the attention weights of edges? 

Furthermore, If alpha can represent the edge_atrr, does it make sense if I just concatenate `x_src`, `x_dst`, and `alpha`, and pass them to MLP to do the edge prediction?

**Reply by rusty1s** [reply]
> Furthermore, If alpha can represent the edge_atrr, does it make sense if I just concatenate x_src, x_dst, and alpha, and pass them to MLP to do the edge prediction?

Yes, I think this is the best option. Let the GNN learn node embeddings and use these for your final edge prediction classifier.

**Reply by haoyuan-l** [reply]
@rusty1s Thank you. I had another confusing problem when I tried to pass the alpha through the GAT layers.
I found the length of `edge_index` increased every time pass through a GAT layer, the same thing happens on `alpha` as well (which should be the same length as `edge_attr`). I turn the `add_self_loops` as `False` when I create the model, so this should not be the reason.

```
print("edge_index_pre")
print(edge_index[:, 0:8])
print(edge_index.size())

out, (edge_index, alpha) = self.conv1(x=x, edge_index=edge_index, edge_attr=edge_attr, return_attention_weights=return_attention_weights)

print("edge_index_aft")
print(edge_index[:, 0:8])
print(edge_index.size())

print("edge_attr")
print(edge_attr.size())

print("alpha")
print(alpha.size())
```
Output:

```
edge_index_pre
tensor([[ 0,  0,  1,  1,  1,  1,  2,  2],
        [ 9,  6,  7,  3,  2, 14, 12,  1]], device='cuda:0')
torch.Size([2, 1076])

edge_index_aft
tensor([[ 0,  0,  1,  1,  1,  1,  2,  2],
        [ 9,  6,  7,  3,  2, 14, 12,  1]], device='cuda:0')  
torch.Size([2, 1447])

edge_attr
tensor([[0.2176],                                            
        [0.2519],                                            
        [0.1010],                                            
        [0.1670],                                            
        [0.2225]], device='cuda:0')                          
torch.Size([1076, 1])  
                                      
alpha                                                        
tensor([[0.0000],                                            
        [0.0000],                                            
        [0.3958],                                            
        [0.0000],                                            
        [0.0000]], device='cuda:0', grad_fn=<SliceBackward0>)
torch.Size([1447, 1])
```

**Reply by haoyuan-l** [reply]
So the problem is, the edge embedding I got from the node embedding is larger than the "actual" edges (as the above output). How can I do edge classification with this?

**Reply by haoyuan-l** [reply]
> print("edge_index_pre")
> print(edge_index[:, 0:8])
> print(edge_index.size())

Sorry for the confusion, I forgot to change the configurations file (`set add_self_loops=False`), and now I have a consistent index for edge_attr and alpha.

**Reply by haoyuan-l** [reply]
> > print("edge_index_pre")
> > print(edge_index[:, 0:8])
> > print(edge_index.size())
> 
> Sorry for the confusion, I forgot to change the configurations file, and now I have a consistent index for edge_attr and alpha.

Apparently, self-loops need to be added to make GAT more effective, is there any way I can remove the attention scores of self-loops before returning them or just use indices operations to get a consistent index? @rusty1s

**Reply by rusty1s** [reply]
If `add_self_loops=False`, you want to apply a skip-connection after the GAT layer instead:
```python
x = conv(x, edge_index) + lin(x)
```
This lets the model preserve central node information.

**Reply by haoyuan-l** [reply]
> If `add_self_loops=False`, you want to apply a skip-connection after the GAT layer instead:
> 
> ```python
> x = conv(x, edge_index) + lin(x)
> ```
> 
> This lets the model preserve central node information.

That's very helpful! Could you explain more about `lin(x)`? Is this a single linear layer with the same `dim_in` and `dim_out`?

**Reply by rusty1s** [reply]
Yes, that is correct. It is known as a "learnable" skip-connection.

**Comment by haoyuan-l** [comment]
I got the training result following the edge embedding  @rusty1s  mentioned above.
![image](https://user-images.githubusercontent.com/91633635/196445498-c9385408-28c8-41f2-8b34-890327d6a60a.png)

The performance of my model is quite poor.
The accuracy/loss of edge and node fluctuates around a value after several epochs. 
(65% for node prediction, 90% for edge prediction)

Is there any way I can improve my model performance, or maybe the GAT is not suitable for this problem?

**Reply by haoyuan-l** [reply]
**Part of the forward func:**

```
out, (edge_index, alpha) = self.GATConv1(x=x, 
                                      edge_index=edge_index, 
                                      edge_attr=edge_attr,
                                      return_attention_weights=return_attention_weights,
                                      )
out = F.dropout(out, p=0.5, training=self.training)
out = out + self.linear(x)

edge_feat = torch.cat([out_src, alpha, out_dst], dim=-1)
edge_out = self.MLP(edge_feat)
```

**Reply by rusty1s** [reply]
Your model looks good. I suggest you add more layers and use non-linearities as well (not sure if you already do this). You might also wanna try to lower your LR or add some form of LR decay.
